commands/ls.go

```diff
@@ -3,142 +3,150 @@ package commands
-					b.err = err
-				}
-				return nil
-			})
-		}(b)
+	if filters.Len() > 0 {
+		var filtered []*builderutil.Builder
+		for _, b := range builders {
+			if filters.Contains("name") && !filters.Match("name", b.NodeGroup.Name) {
+				continue
+			}
+			filtered = append(filtered, b)
+		}
+		builders = filtered
 	}
 
-	if err := eg.Wait(); err != nil {
-		return err
+	if in.quiet {
+		for _, b := range builders {
+			fmt.Println(b.NodeGroup.Name)
+		}
+		return nil
 	}
 
-	currentName := "default"
-	current, err := storeutil.GetCurrentInstance(txn, dockerCli)
-	if err != nil {
+	timeoutCtx, cancel := context.WithTimeout(ctx, 20*time.Second)
+	defer cancel()
+
+	progressMode := progress.PrinterModeAuto
+	if in.format != formatter.FormatTable {
+		progressMode = progress.PrinterModeQuiet
+	}
+	pw := progress.NewPrinter(context.Background(), os.Stderr, progressMode)
+	if err = bc.LoadBuildersData(timeoutCtx, builders, pw); err != nil {
 		return err
 	}
-	if current != nil {
-		currentName = current.Name
-		if current.Name == "default" {
-			currentName = current.Nodes[0].Endpoint
-		}
+	if pwerr := pw.Wait(); err != nil {
+		return pwerr
 	}
 
-	w := tabwriter.NewWriter(os.Stdout, 0, 0, 1, ' ', 0)
-	fmt.Fprintf(w, "NAME/NODE\tDRIVER/ENDPOINT\tSTATUS\tPLATFORMS\n")
-
-	currentSet := false
-	for _, b := range builders {
-		if !currentSet && b.ng.Name == currentName {
-			b.ng.Name += " *"
-			currentSet = true
-		}
-		printngi(w, b)
+	if in.format == formatter.FormatRaw {
+		in.format = lsRawTpl
 	}
 
-	w.Flush()
-
-	return nil
-}
-
-func printngi(w io.Writer, ngi *nginfo) {
-	var err string
-	if ngi.err != nil {
-		err = ngi.err.Error()
-	}
-	fmt.Fprintf(w, "%s\t%s\t%s\t\n", ngi.ng.Name, ngi.ng.Driver, err)
-	if ngi.err == nil {
-		for idx, n := range ngi.ng.Nodes {
-			d := ngi.drivers[idx]
-			var err string
-			if d.err != nil {
-				err = d.err.Error()
-			} else if d.di.Err != nil {
-				err = d.di.Err.Error()
+	fbuilders, hasError := builderutil.FormatBuilders(builders)
+	if err = formatter.Print(fbuilders, in.format, dockerCli.Out(), func(t *tablewriter.Table) {
+		t.SetHeader([]string{"Name/Node", "Driver/Endpoint", "Status", "Platforms"})
+		for _, b := range fbuilders {
+			if b.Current {
+				b.Name += "*"
 			}
-			var status string
-			if d.info != nil {
-				status = d.info.Status.String()
+			if len(b.Err) == 0 {
+				t.Append([]string{aec.Apply(b.Name, aec.Bold), aec.Apply(b.Driver, aec.Bold), "", ""})
+				for _, n := range b.Nodes {
+					t.Append([]string{"  " + n.Name, n.Endpoint, n.Status, strings.Join(n.Platforms, ", ")})
+				}
+			} else {
+				t.Append([]string{aec.Apply(b.Name, aec.Bold), aec.Apply(b.Driver, aec.Bold), "error", ""})
 			}
-			if err != "" {
-				fmt.Fprintf(w, "  %s\t%s\t%s\n", n.Name, n.Endpoint, err)
+		}
+	}); err != nil {
+		return err
+	}
+
+	if hasError && in.format == formatter.FormatTable {
+		_, _ = fmt.Fprintf(dockerCli.Out(), "\n")
+		for _, b := range fbuilders {
```

Comment:
These should be consistently formatted.

```
2 error while checking builder status:
<builder/node>: error message
<builder/node>: error message
```

Again, would prefer to not squash all these different things together.